### DISCUSSION TITLE
Language learning vocab app DB design

### DISCUSSION THREAD

**Discussion by zws33** [zws33/language-app#57] [Ideas]
## Summary

I want my app to be able to generate language learning flash card decks based on tags that a user sets. 

## Requirements:
- store translations
    - they're defined as relationships between two words
- query/group the translations/words by tag. 

## Examples:
- Query all words and their Spanish translations associated with `laundry`.
- Query all English/Italian translation associated with `nightlife`.

## Implementation Details

Here's what I have currently for the DB structure:
```sql
CREATE TABLE language (
    language_code character varying(3) UNIQUE PRIMARY KEY
);

CREATE TABLE word (
    word_id SERIAL PRIMARY KEY,
    word_text character varying(100) NOT NULL UNIQUE,
    language_code character varying(3) NOT NULL REFERENCES language(language_code)
);

CREATE TABLE tag (
    tag_id SERIAL PRIMARY KEY,
    tag_text character varying(500) NOT NULL UNIQUE
);

CREATE TABLE word_tag (
    word_id integer REFERENCES word(word_id),
    tag_id integer REFERENCES tag(tag_id),
    CONSTRAINT word_tag_pkey PRIMARY KEY (word_id, tag_id)
);

CREATE TABLE translation (
    translation_id SERIAL PRIMARY KEY,
    word_id_1 integer NOT NULL REFERENCES word(word_id),
    word_id_2 integer NOT NULL REFERENCES word(word_id)
);
```

## Question

How would I model the tables to support querying words by tags?

**Comment by yousuf-haque** [comment]
You've modeled `words` having `tags` with the table below

> ```sql 
> CREATE TABLE word_tag (
>    word_id integer REFERENCES word(word_id),
>    tag_id integer REFERENCES tag(tag_id),
>    CONSTRAINT word_tag_pkey PRIMARY KEY (word_id, tag_id)
> );
> ```




You can find all words with a given tag:
```sql
SELECT word_id FROM word_tag
WHERE tag_id='laundry';
```

Or find all tags for a given word:
```sql
SELECT tag_id FROM word_tag
WHERE word_id='some_word_id';
```

**Reply by zws33** [reply]
Something I've been debating is should a tag relate to a word or a translation? Currently I have all words in the one table and what language they are is denoted by a column which references a primary key in the `language` table. so as the app grows, that table will contain words in english, spanish, french etc. so getting all words with a given tag will span multiple languages

**Reply by yousuf-haque** [reply]
That's fine, you can join it with the words table, and then add a `AND language_id=whatever` to your `WHERE` clause

**Reply by zws33** [reply]
Also do you have any thoughts on primary key for `word_tag` table? Currently I'm defining it as the composition of `word_id` and `tag_id`. Should the `word_tag` table have an `word_tag_id` column? 

Also, is it redundant for the `tag` table to have an id? I'm not sure if there's a scenario where the tags aren't unique such that I would need another way of identifying them, so I think I could just have the tag by the primary key.

**Reply by yousuf-haque** [reply]
> Also do you have any thoughts on primary key for word_tag table? Currently I'm defining it as the composition of word_id and tag_id

A composite key of those two columns is correct in my opinion. 

> Should the word_tag table have an word_tag_id column?

I don't think this is necessary. Its a relation between two objects, not an object itself. Objects generally need identifiers. 

>  is it redundant for the tag table to have an id?

No. An id is a piece of data to uniquely specify one item in a collection. The `tag_text` is a _display_ value. It's very painful to build an application and find out that you're using a display value to uniquely identify things. 

Also your `unique` constraint on `tag_text` is incorrect - your application cannot handle homonyms. That is one reason not to use display values as unique identifiers.

**Reply by zws33** [reply]
>I don't think this is necessary. Its a relation between two objects, not an object itself. Objects generally need identifiers.

Following this train of thought, would you consider a translation an object or a relationship between two words? Should translation have an ID?
